User saving pop up window for counter and confocal module

lishuo8779 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Improvement of saving function in counter module and confocal module

## Description
<!--- Describe your changes in detail -->
For counter module and confocal module, saving all default data files with a user selected path and user input filename in the pop up window. The user input string would be the prefix for all data files. The default saving folder has been deactivated in this process and only keeps log files.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
By default Qudi saves data in the unchangeable folder (\Data) with timestamps as prefix. It would be convenient for the users to name and save data in their own folder.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Tested with the dummy modules (confocal & counter) on a MacOS machine (Catalina v10.15.2)

## Screenshots (only if appropriate, delete if not):

## Types of changes
<!--- What types of changes does your code introduce? Put an 'x' in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an 'x' in all the boxes that apply. -->
<!--- If you're unsure about any of these, ask. -->
- [x] My code follows the code style of this project.
- [ ] I have documented my changes in the changelog (`documentation/changelog.md`)
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added/updated for the module the config example in the docstring of the class accordingly.
- [x] I have checked that the change does not contain obvious errors (syntax, indentation, mutable default values).
- [x] I have tested my changes using 'Load all modules' on the default dummy configuration with my changes included.
- [ ] All changed Jupyter notebooks have been stripped of their output cells.

lishuo8779 committed:
Saving file names from the user with pop up window

lishuo8779 committed:
Saving file names from the user with pop up window for counter module

lishuo8779 committed:
Canceled saving raw png file in confocalgui.py because it looks redundant

lishuo8779 committed:
Merge pull request #5 from lishuo8779/user_saving

User saving

MdevriesPhys reviewed (dismissed):
Looks like it's all good :)

lishuo8779 commented:
Received AttributeError when loading all modules. Considering fix.
<img width="515" alt="Screen Shot 2020-02-27 at 3 11 53 pm" src="https://user-images.githubusercontent.com/12246989/75411496-8a9ede80-5973-11ea-8e8e-9b0cc2b959b7.png">

lishuo8779 committed:
Fixed the AttributeError warning when loading all modules in last version.

marcocapelli commented:
Errors:
- If I change my mind and press Cancel on the pop-up window it raises an error: TypeError: 'int' object is not iterable.
- The last 4 characters of filename I choose is cut in the real file. If I save "test2" it is saved as "t", if I try to save "abcdefgh" it is saved as "abcd" + all the suffix as expected.
- Still saving the raw png files when the "save raw png" checkbox is not checked.
- If the file already exists it does not recognize that and instead overwrites it with the new file.
- When going back to the automatic saving behaviour the confocal map saving gives an error: the filename variable is None and the error raises when trying to get filename[:4]
- When going back to the automatic saving behaviour, there is a typo in the counter saving function.

Personal comments:
- I still do not like that images are saved with an additional _fig at the end of the file name.
- You do not need to keep the old lines commented, you can just delete them. If they are needed again, the version control of git can bring them back.
- Ulm made changes to the saving logic as well! We should rebase our master before applying modification to see if anything new is conflicting.

lishuo8779 committed:
1. Fixed the pop-up Error window when user presses the Cancel button during saving.
2. Fixed the bug that the last 4 characters in user defined filename get cut in Linux.
3. Fixed the error when switch back to default saving mode.
4. Add saving .tiff file in save_logic
5. Add an empty .txt file in confocal saving process as a overwrite indicator when user inputs duplicated file names.
6. Changed the postfix of filename.

lishuo8779 commented:
> Errors:
> 
> * If I change my mind and press Cancel on the pop-up window it raises an error: TypeError: 'int' object is not iterable.

Fixed.

> * The last 4 characters of filename I choose is cut in the real file. If I save "test2" it is saved as "t", if I try to save "abcdefgh" it is saved as "abcd" + all the suffix as expected.

Should be fixed. Please test on Linux.

> * Still saving the raw png files when the "save raw png" checkbox is not checked.

The raw png file comes from the default saving process, just named in "raw" way. The checkbox controls another weird png file which is not saved anymore. Need some help to remove the checkbox.

> * If the file already exists it does not recognize that and instead overwrites it with the new file.

The files are saved with postfix so the dialog window can't check if the user input something different from last input. I add an empty txt file naming directly with the user input string as an indicator file, but this is really inelegant... Any creative thoughts? To get a system warning window but without redundant file saved? 

> * When going back to the automatic saving behaviour the confocal map saving gives an error: the filename variable is None and the error raises when trying to get filename[:4]

Should be fixed. Please test on Linux.

> * When going back to the automatic saving behaviour, there is a typo in the counter saving function.

Fixed.
> 
> Personal comments:
> 
> * I still do not like that images are saved with an additional _fig at the end of the file name.

Fine fine... I surrender... for a group of picky users... the _fig postfix is deleted now.

> * You do not need to keep the old lines commented, you can just delete them. If they are needed again, the version control of git can bring them back.

Old lines deleted.

> * Ulm made changes to the saving logic as well! We should rebase our master before applying modification to see if anything new is conflicting.

Sounds great. Going to merge into the user_saving branch to test.

MdevriesPhys commented:
Okay so I think I'm going mad, I can't seem to get any of the options for automated saving to work... (or even find the settings for it). Help?

marcocapelli commented:
If you mean that you want to check the default Ulm-QuDi system way of
saving, you have to add in the config file the option (under the save_logic
section):
save_into_default_directory: True

As the code is written, if the option is not specified the default value is
False and the user is prompted to insert their own file name in the
dialogue box.

On Thu, 9 Apr 2020 at 12:41, Mitchell de Vries <notifications@github.com>
wrote:

> Okay so I think I'm going mad, I can't seem to get any of the options for
> automated saving to work... (or even find the settings for it). Help?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CNBP-RMIT/qudi/pull/6#issuecomment-611296129>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF6W2C4XNYW3DBPUUELDMU3RLUYU7ANCNFSM4K36YN6Q>
> .
>


-- 
Marco Capelli
Research Assistant
----
School of Science, RMIT University
Melbourne, 3001 Victoria, Australia
----
ORCID <https://orcid.org/0000-0002-8353-8492> - Google Scholar
<https://scholar.google.com/citations?pli=1&authuser=3&user=4WBMn_YAAAAJ> -
LinkedIn <https://www.linkedin.com/in/marco-capelli/>

MdevriesPhys commented:
I meant saving after each loop of a scan?

marcocapelli commented:
Oh, I see, a new functionality.
The scan never knows it stopped really (if I remember correctly). What
happens is that any stopping interaction sends the "stop requested" signal
to the scanning logic. Next time the scanning is moving to the next line,
if the stop was requested it will just not proceed and instead will exit
the function.
What I did when I needed to know that the scanning finished (to restart the
counter, for example) was to add the behaviour is the close_scanner()
method of the tilt_interfuse class (which is also called by the optimizer).
Otherwise, there are the stop_scanning signal and method in the
confocal_logic.


On Thu, 9 Apr 2020 at 14:20, Mitchell de Vries <notifications@github.com>
wrote:

> I meant saving after each loop of a scan?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CNBP-RMIT/qudi/pull/6#issuecomment-611319229>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF6W2C3J3FL3XBAA25CEQL3RLVEGJANCNFSM4K36YN6Q>
> .
>


-- 
Marco Capelli
Research Assistant
----
School of Science, RMIT University
Melbourne, 3001 Victoria, Australia
----
ORCID <https://orcid.org/0000-0002-8353-8492> - Google Scholar
<https://scholar.google.com/citations?pli=1&authuser=3&user=4WBMn_YAAAAJ> -
LinkedIn <https://www.linkedin.com/in/marco-capelli/>

MdevriesPhys commented:
Yeah okay, that makes sense, looks like it's all behaving right on for me then :)

lishuo8779 commented:
> Yeah okay, that makes sense, looks like it's all behaving right on for me then :)

Yipee! Finally passed 😁 Seems the picky peer-review exists not only in paper submission...

marcocapelli committed:
The "Save raw" checkbox is not used anymore in the code. Removed from the UI and cleaned up the relevant code in the gui python file.

marcocapelli commented:
The bad cop is back!
All works well and I doubled check the errors from last time, they are all solved.
I also removed the checkbox that has no functionality anymore.

Only one major thing is missing before merging: our master has been updated to Ulm's version (a while ago now, we should do it again soon). This new version has conflicts with this branch that must be solved before merging.

Additional notes: (these are meant as future works, as it works and we can merge anyway)
- We should re-think the way it checks for existing files. This works but also creates redundant additional files.
- Tiff files should not have all the formatting as the png files have. The idea would be to replicate the kind of file you expect to be saved by a commercial microscope, which is a matrix of intensity values at each pixel. This file type is easily opened by software like ImageJ for image analysis. Also, tiff files can be saved with 3D matrices where each channel is a different layer, which would be extremely helpful to avoid thousands of files saved when there are multiple channels. My suggestion is to go and check the scikit-image python package.
- Because of what discussed above, a tiff file should be saved only if we are trying to save 2D/3D images. The counter data should not have a tiff file associated.
- I still not like the fact that the program asks me to input a file name and then it freely adds a lot more words to the filename. Based on all other systems/software I used in our labs and across RMIT, if the program asks me a name that name will be used for the final file. Unless strictly required to avoid saving 2 or more datasets with the same name and extension (i.e. LightField when you ask to save each frame as separate a file. Of course, it is not a groundbreaking option, and users can easily deal with it, but it is an unusual behavior when compared to commercial software.

marcocapelli reviewed (approved):
Comments specified above ^

lishuo8779 committed:
Merge branch 'master' into user_saving
